src/lib/ares_process.c

```diff
@@ -816,6 +854,77 @@ static struct server_state *ares__random_server(ares_channel_t *channel)
   return NULL;
 }
 
+/* Pick a server from the list with failover behavior.
+ *
+ * We default to using the first server in the sorted list of servers. That is
+ * the server with the lowest number of consecutive failures, and then the
+ * highest priority server (by idx) if there is a draw.
+ *
+ * However, this by itself has two issues:
+ * (1) If a server temporarily goes down and hits some failures, then that
+ *     server will never be retried until all other servers hit the same number
+ *     of failures. This may prevent the server from being retried for a long
+ *     time.
+ * (2) If all servers are failed, then we will continue to select the
+ *     server with the lowest number of consecutive failures even though it may
+ *     be more broken than other servers later in the list.
+ *
+ * For issue (1), with some probability we select a failed server to retry
+ * instead.
+ *
+ * For issue (2), if all servers in the list have a serious failure, then we
+ * fall back to selecting a random server.
+ */
+static struct server_state *ares__failover_server(ares_channel_t *channel)
+{
+  struct server_state *server;
+
+  /* Select the best choice of server, based on consecutive failures and
+   * server idx.
+   */
+  server = ares__slist_first_val(channel->servers);
+
+  /* Defensive code against no servers being available on the channel. */
+  if (server == NULL) {
+    return NULL;
+  }
+
+  if (channel->server_serious_fail_limit > 0 &&
+      server->consec_failures >= channel->server_serious_fail_limit) {
+    /* All servers have a serious failure so fall back to a random server. */
+    return ares__random_server(channel);
+  }
+
+  if (channel->server_retry_chance > 0) {
+    /* Generate a random byte to decide whether to retry a failed server. */
+    unsigned char c;
+    ares__rand_bytes(channel->rand_state, &c, 1);
+    if (c % channel->server_retry_chance == 0) {
+      /* Select a failed server to retry. */
+      ares__slist_node_t  *node;
+      for (node = ares__slist_node_first(channel->servers); node != NULL;
+           node = ares__slist_node_next(node)) {
+        struct server_state *node_val = ares__slist_node_val(node);
+        if (node_val != NULL && node_val->consec_failures > 0) {
+          /* Check that the next retry time for failed servers has passed. If
+           * not then break out and return the best server.
+           */
+          struct timeval now = ares__tvnow();
+          if (ares__timedout(&now, &channel->server_next_retry_time)) {
+            timeadd(&now, channel->server_retry_delay);
+            channel->server_next_retry_time = now;
+            return node_val;
+          }
+          break;
```

Comment:
Wouldn't it be better to check the timeout outside of the loop and skip it altogether?

Also we can probably check the last server in the list to see if we have any failed servers to retry.

Reply:
I put the timeout check inside the loop because I wanted to avoid the call to `ares_tvnow()` if there were no failed servers to retry (optimising the mainline "all servers work" case and avoiding system calls).

But yep, a simple check of the last server (can use `ares__slist_node_last()`) would improve this. I will fix this. Thanks!

Reply:
at least on linux ares_tvnow() should be very fast due to it using a vdso call so I wouldn't really worry about optimizing out that call.  But otherwise, yeah, checking the last member to see if any servers have failed makes sense due to sorting.